lightstep/trace_context.py

```diff
@@ -0,0 +1,311 @@
+    There is a set of test cases defined for testing any implementation of the
+    document. This set of test cases can be found here:
+    https://github.com/w3c/trace-context/tree/master/test
+
+    This set of test cases will be referred to as "the test suite" in the
+    comments of this implementation.
+    """
+
+    def inject(self, span_context, carrier):
+
+        carrier[_TRACEPARENT] = "00-{}-{}-{}".format(
+            format(span_context.trace_id, "032x"),
+            format(span_context.span_id, "016x"),
+            format(span_context.baggage.pop("trace-flags", 0), "02x")
+        )
+
+        carrier.update(span_context.baggage)
+
+    def extract(self, carrier):
+        traceparent = None
+        tracestate = None
+
+        multiple_header_template = "Found more than one header value for {}"
+
+        # FIXME Define more specific exceptions and use them instead of the
+        # generic ones that are being raised below
+
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#header-name
+
+        trace_context_free_carrier = {}
+
+        for key, value in carrier.items():
+
+            lower_key = key.lower()
+
+            # The document requires that the headers be accepted regardless of
+            # the case of their characters. This means that the keys of carrier
+            # may contain 2 or more strings that match traceparent or
+            # tracestate when the case of the characters of such strings is
+            # ignored. The document does not specify what is to be done in such
+            # a situation, this implementation will raise an exception.
+            if lower_key == _TRACEPARENT:
+
+                if traceparent is None:
+                    traceparent = value
+
+                else:
+                    raise Exception(
+                        multiple_header_template.format(_TRACEPARENT)
+                    )
+
+            elif lower_key == _TRACESTATE:
+
+                if tracestate is None:
+                    tracestate = value
+
+                else:
+                    raise Exception(
+                        multiple_header_template.format(_TRACEPARENT)
+                    )
+
+            else:
+                trace_context_free_carrier[key] = value
+
+        if traceparent is None:
+            # https://www.w3.org/TR/trace-context/#no-traceparent-received
+            _LOG.warning("No traceparent was received")
+            return SpanContext(
+                trace_id=getrandbits(128), span_id=getrandbits(64)
+            )
+
+        else:
+
+            version_match = _VERSION.match(traceparent)
+
+            if version_match is None:
+                # https://www.w3.org/TR/2019/CR-trace-context-20190813/#versioning-of-traceparent
+                _LOG.warning(
+                    "Unable to parse version from traceparent"
+                )
+                return SpanContext(
+                    trace_id=getrandbits(128), span_id=getrandbits(64)
```

Comment:
Are we cool here using 128-bits for trace id?

Reply:
The document [specifies](https://www.w3.org/TR/trace-context/#version-format) 16 bytes for `trace-id`. Am I understanding this wrong? :thinking:

Reply:
Sure, we need to use 16 bytes for the format (and propagate 16 bytes ids as well). But I'm not sure we properly handle such a long id in the rest of the code base (`SpanContext` within `lightstep` module), nor the backend. 

I.e. we should do what we did for B3, where we propagate the 'original' id but internally use only 8 bytes ;)

Reply:
@codeboten what do you think?

Reply:
@carlosalberto after discussing this with @codeboten I have submitted a commit that truncates the value of `trace_id` automatically when exiting a span context. I think this is the right approach, since we should keep the span context attributes as they are for `inject` and `extract` and only do truncation for LightStep specific code.